File: backend/src/main/java/com/alexandria/service/BibliotecaService.java
package com.alexandria.service;

import com.alexandria.dto.biblioteca.AdicionarLivroRequest;
import com.alexandria.dto.biblioteca.ItemBibliotecaResponse;
import com.alexandria.dto.livro.LivroResumoResponse;
import com.alexandria.exception.DuplicateResourceException;
import com.alexandria.exception.ResourceNotFoundException;
import com.alexandria.model.ItemBiblioteca;
import com.alexandria.model.Livro;
import com.alexandria.model.StatusLeitura;
import com.alexandria.model.User;
import com.alexandria.repository.ItemBibliotecaRepository;
import com.alexandria.repository.UserRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
public class BibliotecaService {

    private final ItemBibliotecaRepository itemBibliotecaRepository;
    private final UserRepository userRepository;
    private final LivroService livroService;
    private final GamificacaoService gamificacaoService;

    public BibliotecaService(ItemBibliotecaRepository itemBibliotecaRepository,
                              UserRepository userRepository,
                              LivroService livroService,
                              GamificacaoService gamificacaoService) {
        this.itemBibliotecaRepository = itemBibliotecaRepository;
        this.userRepository = userRepository;
        this.livroService = livroService;
        this.gamificacaoService = gamificacaoService;
    }

    @Transactional
    public ItemBibliotecaResponse adicionar(Long usuarioId, AdicionarLivroRequest request) {
        Livro livro = livroService.obterOuCriarLivroLocal(request.identificadorExterno());

        if (itemBibliotecaRepository.existsByUsuarioIdAndLivroId(usuarioId, livro.getId())) {
            throw new DuplicateResourceException("Este livro já está na sua biblioteca.");
        }

        User usuario = userRepository.getReferenceById(usuarioId);
        StatusLeitura status = request.statusLeitura() == null ? StatusLeitura.QUERO_LER : request.statusLeitura();

        ItemBiblioteca item = ItemBiblioteca.builder()
                .usuario(usuario)
                .livro(livro)
                .statusLeitura(status)
                .build();
        item = itemBibliotecaRepository.save(item);

        gamificacaoService.livroAdicionado(usuarioId, status);

        return paraResponse(item);
    }

    public List<ItemBibliotecaResponse> listar(Long usuarioId, StatusLeitura filtroStatus, Boolean favorito) {
        return itemBibliotecaRepository.findByUsuarioIdOrderByCriadoEmDesc(usuarioId).stream()
                .filter(item -> filtroStatus == null || item.getStatusLeitura() == filtroStatus)
                .filter(item -> favorito == null || item.isFavorito() == favorito)
                .map(this::paraResponse)
                .toList();
    }

    @Transactional
    public ItemBibliotecaResponse atualizarStatus(Long usuarioId, Long itemId, StatusLeitura novoStatus) {
        ItemBiblioteca item = buscarItem(usuarioId, itemId);
        StatusLeitura statusAntigo = item.getStatusLeitura();
        item.setStatusLeitura(novoStatus);

        gamificacaoService.statusAlterado(usuarioId, statusAntigo, novoStatus);

        return paraResponse(item);
    }

    @Transactional
    public ItemBibliotecaResponse alternarFavorito(Long usuarioId, Long itemId) {
        ItemBiblioteca item = buscarItem(usuarioId, itemId);
        item.setFavorito(!item.isFavorito());

        gamificacaoService.favoritoAlternado(usuarioId, item.isFavorito());

        return paraResponse(item);
    }

    @Transactional
    public void remover(Long usuarioId, Long itemId) {
        ItemBiblioteca item = buscarItem(usuarioId, itemId);
        itemBibliotecaRepository.delete(item);
        gamificacaoService.livroRemovido(usuarioId, item.getStatusLeitura(), item.isFavorito());
    }

    private ItemBiblioteca buscarItem(Long usuarioId, Long itemId) {
        return itemBibliotecaRepository.findByIdAndUsuarioId(itemId, usuarioId)
                .orElseThrow(() -> new ResourceNotFoundException("Item da biblioteca não encontrado."));
    }

    private ItemBibliotecaResponse paraResponse(ItemBiblioteca item) {
        Livro livro = item.getLivro();
        LivroResumoResponse livroResumo = new LivroResumoResponse(
                livro.getIdentificadorExterno(),
                livro.getTitulo(),
                livro.getAutor(),
                livro.getCapa(),
                livro.getCategoria(),
                livro.getDataPublicacao()
        );
        return new ItemBibliotecaResponse(
                item.getId(),
                livroResumo,
                item.getStatusLeitura(),
                item.isFavorito(),
                item.getCriadoEm(),
                item.getAtualizadoEm()
        );
    }
}

File: backend/src/services/bibliotecaService.js
const prisma = require('../lib/prisma')
const livroService = require('./livroService')
const gamificacaoService = require('./gamificacaoService')
const { DuplicateError, NotFoundError } = require('../lib/errors')

function paraResponse(item) {
  return {
    id: item.id,
    livro: {
      identificadorExterno: item.livro.identificadorExterno,
      titulo: item.livro.titulo,
      autor: item.livro.autor,
      capa: item.livro.capa,
      categoria: item.livro.categoria,
      dataPublicacao: item.livro.dataPublicacao,
    },
    statusLeitura: item.statusLeitura,
    favorito: item.favorito,
    criadoEm: item.criadoEm,
    atualizadoEm: item.atualizadoEm,
  }
}

async function adicionar(usuarioId, { identificadorExterno, statusLeitura }) {
  const livro = await livroService.obterOuCriarLivroLocal(identificadorExterno)

  const existente = await prisma.itemBiblioteca.findUnique({
    where: { usuarioId_livroId: { usuarioId, livroId: livro.id } },
  })
  if (existente) {
    throw new DuplicateError('Este livro já está na sua biblioteca.')
  }

  const status = statusLeitura ?? 'QUERO_LER'
  const item = await prisma.itemBiblioteca.create({
    data: { usuarioId, livroId: livro.id, statusLeitura: status },
    include: { livro: true },
  })

  await gamificacaoService.livroAdicionado(usuarioId, status)

  return paraResponse(item)
}

async function listar(usuarioId, { status, favorito }) {
  const itens = await prisma.itemBiblioteca.findMany({
    where: {
      usuarioId,
      ...(status ? { statusLeitura: status } : {}),
      ...(favorito !== undefined ? { favorito } : {}),
    },
    include: { livro: true },
    orderBy: { criadoEm: 'desc' },
  })
  return itens.map(paraResponse)
}

async function atualizarStatus(usuarioId, itemId, novoStatus) {
  const item = await buscarItem(usuarioId, itemId)
  const statusAntigo = item.statusLeitura

  const atualizado = await prisma.itemBiblioteca.update({
    where: { id: item.id },
    data: { statusLeitura: novoStatus },
    include: { livro: true },
  })

  await gamificacaoService.statusAlterado(usuarioId, statusAntigo, novoStatus)

  return paraResponse(atualizado)
}

async function alternarFavorito(usuarioId, itemId) {
  const item = await buscarItem(usuarioId, itemId)

  const atualizado = await prisma.itemBiblioteca.update({
    where: { id: item.id },
    data: { favorito: !item.favorito },
    include: { livro: true },
  })

  await gamificacaoService.favoritoAlternado(usuarioId, atualizado.favorito)

  return paraResponse(atualizado)
}

async function remover(usuarioId, itemId) {
  const item = await buscarItem(usuarioId, itemId)
  await prisma.itemBiblioteca.delete({ where: { id: item.id } })
  await gamificacaoService.livroRemovido(usuarioId, item.statusLeitura, item.favorito)
}

async function buscarItem(usuarioId, itemId) {
  const item = await prisma.itemBiblioteca.findFirst({ where: { id: itemId, usuarioId } })
  if (!item) {
    throw new NotFoundError('Item da biblioteca não encontrado.')
  }
  return item
}

module.exports = { adicionar, listar, atualizarStatus, alternarFavorito, remover }
